components/sync15/src/client.rs

```diff
@@ -48,27 +54,61 @@ pub enum Sync15ClientResponse<T> {
+    {
+        let route: String = resp.url.path().into();
+        Ok(if resp.is_success() {
+            let record: T = resp.json()?;
+            let last_modified = resp
+                .headers
+                .get(header_names::X_LAST_MODIFIED)
+                .and_then(|s| ServerTimestamp::from_str(s).ok())
+                .ok_or_else(|| ErrorKind::MissingServerTimestamp)?;
+            Sync15ClientResponse::Success {
+                status: resp.status,
+                record,
+                last_modified,
+                route,
+            }
+        } else {
+            let status = resp.status;
+            match status {
+                404 => Sync15ClientResponse::NotFound { route },
+                401 => Sync15ClientResponse::Unauthorized { route },
+                412 => Sync15ClientResponse::PreconditionFailed { route },
+                // / TODO: 5XX errors should parse backoff etc.
+                500..=600 => Sync15ClientResponse::ServerError { route, status },
+                _ => Sync15ClientResponse::RequestFailed { route, status },
+            }
+        })
+    }
+
     pub fn create_storage_error(self) -> ErrorKind {
-        let (code, route): (u16, String) = match self {
-            Sync15ClientResponse::Success { route, .. } => {
+        let inner = match self {
+            Sync15ClientResponse::Success { status, route, .. } => {
                 // This should never happen as callers are expected to have
                 // already special-cased this response, so warn if it does.
                 // (or maybe we could panic?)
                 log::warn!("Converting success response into an error");
-                (200, route)
+                StorageHttpError::RequestFailed { status, route }
+            }
+            Sync15ClientResponse::NotFound { route, .. } => StorageHttpError::NotFound { route },
+            Sync15ClientResponse::Unauthorized { route, .. } => {
+                StorageHttpError::Unauthorized { route }
+            }
+            Sync15ClientResponse::ServerError { status, route } => {
+                StorageHttpError::ServerError { status, route }
+            }
+            Sync15ClientResponse::RequestFailed { status, route } => {
+                StorageHttpError::RequestFailed { status, route }
+            }
+            Sync15ClientResponse::PreconditionFailed { route } => {
+                StorageHttpError::PreconditionFailed { route }
             }
-            Sync15ClientResponse::NotFound { route, .. } => (404, route),
-            Sync15ClientResponse::Unauthorized { route, .. } => (401, route),
-            Sync15ClientResponse::ServerError { status, route } => (status, route),
-            Sync15ClientResponse::RequestFailed { status, route } => (status, route),
         };
-        ErrorKind::StorageHttpError {
-            code,
-            route: route.clone(),
-        }
+        ErrorKind::StorageHttpError(inner)
```

Comment:
Could we use the `Sync15ClientResponse` directly as `inner` here? It seems like there's a lot of overlap between it and `StorageHttpError` now, save for special-casing success.

Reply:
> Could we use the `Sync15ClientResponse` directly as `inner` here? It seems like there's a lot of overlap between it and `StorageHttpError` now, save for special-casing success.

I tried this and it got tricky fairly quickly - as you mention `Success` is a bit tricky, and the fact `Sync15ClientResponse` is generic made it a real pain (eg, I tried using `()` as the `T`, but it's still a different type, leading to code like [this](https://github.com/mozilla/application-services/blob/master/components/sync15/src/client.rs#L118). I might have another stab though.

Reply:
The second stab came out much better, so yeah, thanks for pushing back on my laziness ;)